crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1101 @@
+        Item = (
+            String,
+            Option<SmallVec<[qiskit_circuit::operations::Param; 3]>>,
+            SmallVec<[PhysicalQubit; 2]>,
+        ),
+    >,
+    target: &Target,
+) -> f64 {
+    let mut gate_fidelities = Vec::new();
+    let mut score_instruction =
+        |inst_name: &str,
+         inst_params: &Option<SmallVec<[Param; 3]>>,
+         inst_qubits: &SmallVec<[PhysicalQubit; 2]>| {
+            if let Ok(names) = target.operation_names_for_qargs(Some(inst_qubits)) {
+                for name in names {
+                    if let Ok(target_op) = target.operation_from_name(name) {
+                        let are_params_close = if let Some(params) = inst_params {
+                            params.iter().zip(target_op.params.iter()).all(|(p1, p2)| {
+                                p1.is_close(py, p2, 1e-10)
+                                    .expect("Unexpected parameter expression error.")
+                            })
+                        } else {
+                            false
+                        };
+                        let is_parametrized = target_op
+                            .params
+                            .iter()
+                            .any(|param| matches!(param, Param::ParameterExpression(_)));
+                        if target_op.operation.name() == inst_name
+                            && (is_parametrized || are_params_close)
+                        {
+                            match target[name].get(Some(inst_qubits)) {
+                                Some(Some(props)) => {
+                                    gate_fidelities.push(1.0 - props.error.unwrap_or(0.0))
+                                }
+                                _ => gate_fidelities.push(1.0),
+                            }
+                            break;
+                        }
+                    }
+                }
+            }
+        };
+
+    for (inst_name, inst_params, inst_qubits) in synth_circuit {
+        score_instruction(&inst_name, &inst_params, &inst_qubits);
+    }
+    1.0 - gate_fidelities.into_iter().product::<f64>()
+}
+
+// This is the outer-most run function. It is meant to be called from Python
+// in `UnitarySynthesis.run()`.
+#[pyfunction]
+#[pyo3(name = "run_default_main_loop")]
+fn py_run_main_loop(
+    py: Python,
+    dag: &mut DAGCircuit,
+    qubit_indices: &Bound<'_, PyList>,
+    min_qubits: usize,
+    target: &Target,
+    coupling_edges: &Bound<'_, PyList>,
+    approximation_degree: Option<f64>,
+    natural_direction: Option<bool>,
+) -> PyResult<DAGCircuit> {
+    let dag_to_circuit = imports::DAG_TO_CIRCUIT.get_bound(py);
+
+    // Run synthesis recursively over control flow ops, mapping qubit indices
+    let node_ids: Vec<NodeIndex> = dag.op_nodes(false).collect();
+    for node in node_ids {
+        let NodeType::Operation(inst) = &dag.dag()[node] else {
+            panic!("DAG node must be an instruction")
+        };
```

Comment:
You can use `unwrap_operation` here too.
```suggestion
        let inst = &dag.dag()[node].unwrap_operation();
```